src/utils/happo.js

```diff
@@ -0,0 +1,48 @@
+const React = require('react');
+const { LiveProvider, LivePreview } = require('react-live');
+
+const { ThemeProvider } = require('../themes');
+
+// Some components have dependencies that make them hard to render statically.
+// For now, we can just exclude these from the visual regression test suite.
+const excludedComponents = ['Popover', 'Dropdown'];
```

Comment:
Likely. It’s definitely a trade-off thing. I might make it configurable per project, or switch to Chrome (headless) for the initial render.

Reply:
Yeah, headless Chrome in particular, is what I was wondering about.

Reply:
I'll investigate this at some point going forward. The jsdom environment is a little limiting. Not as bad as I first thought (they've clearly did a good job adding support for all kinds of dom operations). Headless chrome is a good replacement candidate. The reason I went with jsdom to begin with is that it has fewer dependencies. Headless chrome is surprisingly tricky to install and run (I spent about a week setting it up for the chrome worker instances on happo.io). Having a simple local dev runner (`happo dev --only Button`) is an important goal for me, and requiring users to install headless chrome could make that goal harder. 

Anywho, I'm ranting here. I appreciate the feedback. 👍

Reply:
The rendering limitations of `jsdom` are not entirely unexpected and are certainly not a blocker for us.

I am hopeful that at some point, we can just come up with an appropriate mock and/or polyfill for the bits of code that it is choking on just enough to get things to render, maybe not be interactive, and thus glean the value of these VR tests for our Popup/Dropdown/Select/Tooltip/etc components that currently rely on actual DOM.

Reply:
Good to know. 

In general, I have mixed feelings about libraries like popper.js (which is what caused me to exclude Popover and Dropdown). They solve a number of edge-cases with pop-up elements, but usually at the cost of simplicity. Did you consider writing something simpler yourself? It seems popper.js adds about 20KB (minified, not gzipped) to the bundle. Even though that's not a whole lot, it could be significant for some deployments. 

I actually went down the polyfill route for a while, trying to do just enough to make popper.js happy. But one thing that was really hard to polyfill was `document.createRange`, which jsdom has [a 7 year old issue](https://github.com/tmpvar/jsdom/issues/317) open to add support for. I failed to even understand how/why it was used, and even more so trying to implement a mock for it...

Reply:
I couldn't agree more.  I have the same mixed feelings.  We did consider writing all of our own popover positioning logic, and in fact did that on an internal component library last year, which only uses fixed positioning. Way simpler and works well for CA Agile Central, I was just unsure if it would fit everyones different use cases out in the open.  It is a really a matter of time and where we should be spending it.

I tried the same polyfill for Popper too, and saw the same ancient jsdom issue.  I had a working simple `createRange` stub/polyfill, but that just led to another DOM issue. Resultantly, I opted to mock Popper for our unit tests instead.  See https://github.com/FezVrasta/popper.js/issues/478

Reply:
18KB unzipped isn't big at all, don't feel bad about it. 😄 

I wouldn't hesitate adding something like popper.js to an application bundle myself, but since material-ui is a library it's probably good to be cautious about adding too much weight. I would assume that most material-ui consumers would use only a fraction of popper.js in their applications. That being said, I don't know much about writing positioning engines.

Reply:
Actually the heaviest part of the library is what is needed by the average user, the additional features that can be removed add just few extra kbs.

Reply:
Ah, that's good to know!